beluga/include/beluga/algorithm/amcl_core.hpp

```diff
@@ -0,0 +1,240 @@
+      spatial_hasher_type spatial_hasher,
+      const AmclParams& params = AmclParams{},
+      ExecutionPolicy execution_policy = std::execution::seq)
+      : params_{params},
+        motion_model_{std::move(motion_model)},
+        sensor_model_{std::move(sensor_model)},
+        execution_policy_{std::move(execution_policy)},
+        spatial_hasher_{std::move(spatial_hasher)},
+        random_probability_estimator_{params_.alpha_slow, params_.alpha_fast},
+        update_policy_{beluga::policies::on_motion(params_.update_min_d, params_.update_min_a)},
+        resample_policy_{beluga::policies::every_n(params_.resample_interval)},
+        random_state_generator_(std::move(random_state_generator)) {
+    if (params_.selective_resampling) {
+      resample_policy_ = resample_policy_ && beluga::policies::on_effective_size_drop;
+    }
+  }
+
+  /// Returns a reference to the current set of particles.
+  [[nodiscard]] const auto& particles() const { return particles_; }
+
+  /// Initialize particles using a custom distribution.
+  template <class Distribution>
+  void initialize(Distribution distribution) {
+    particles_ = beluga::views::sample(std::move(distribution)) |                    //
+                 ranges::views::transform(beluga::make_from_state<particle_type>) |  //
+                 ranges::views::take_exactly(params_.max_particles) |                //
+                 ranges::to<beluga::TupleVector>;
+    force_update_ = true;
+  }
+
+  /// Initialize particles with a given pose and covariance.
+  /**
+   * \tparam CovarianceT type representing a covariance, compliant with state_type.
+   * \throw std::runtime_error If the provided covariance is invalid.
+   */
+  template <class CovarianceT>
+  void initialize(state_type pose, CovarianceT covariance) {
+    initialize(beluga::MultivariateNormalDistribution{pose, covariance});
+  }
+
+  /// Update the map used for localization.
+  void update_map(map_type map) { sensor_model_.update_map(std::move(map)); }
+
+  /// Update particles based on motion and sensor information.
+  /**
+   * This method performs a particle filter update step using motion and sensor data. It evaluates whether
+   * an update is necessary based on the configured update policy and the force_update flag. If an update
+   * is required, the motion model and sensor model updates are applied to the particles, and the particle
+   * weights are adjusted accordingly. Also, according to the configured resampling policy, the particles
+   * are resampled to maintain diversity and prevent degeneracy.
+   *
+   * \param state_at_measurement State at the measurement time.
+   * \param measurement Measurement data.
+   * \return An optional pair containing the estimated pose and covariance after the update,
+   *         or std::nullopt if no update was performed.
+   */
+  auto update(state_type state_at_measurement, measurement_type measurement) -> std::optional<estimation_type> {
```

Comment:
@serraramiro1 meta: hmm, this should be taking a control action like generic motion models do. Using a window of state measurements is something specific to a kind of motion models. But pulling the control window out is inconvenient for the user, so perhaps this class should be specialized to include the window when a specific kind of motion model is in use. 

That said, I'm OK with waiting until we have a motion model that doesn't use a control window to implement that.

Reply:
I see, I think I got confused by this taking in a `state_type` instead of what seems to be a more generic `control_type`.
DYT we should exclude the rolling window here as it seems model-specific and take in a `control_type`?

Reply:
Yes but it's kinda inconvenient because then you have to store the window elsewhere. The nice thing would be to have a specialization that includes the window for the motion model that use one. It's some work though. I won't argue against leaving the TODO behind.

Reply:
Also just noticed that the `on_motion` policy needs a state (at least for the current motion models). So even if we pass a `control_type` we'd need a way to extract the pose out of it :-1:

Reply:
Well, `on_motion` could integrate control actions or AMCL could track prior state estimates. There is room for generalization, but we don't have to generalize now.

Reply:
Done!